airflow/operators/mysql_to_s3_operator.py

```diff
@@ -0,0 +1,117 @@
+    :param query: the sql query to be executed.
+    :type query: str
+    :param s3_bucket: bucket where the data will be stored
+    :type s3_bucket: str
+    :param s3_key: desired key for the file. It includes the name of the file.
+        If a csv file is wanted, the param must end with ".csv".
+    :type s3_key: str
+    :param mysql_conn_id: reference to a specific mysql database
+    :type mysql_conn_id: str
+     :param aws_conn_id: reference to a specific S3 connection
+    :type aws_conn_id: str
+    :param verify: Whether or not to verify SSL certificates for S3 connection.
+        By default SSL certificates are verified.
+        You can provide the following values:
+
+        - False: do not validate SSL certificates. SSL will still be used
+                 (unless use_ssl is False), but SSL certificates will not be
+                 verified.
+        - path/to/cert/bundle.pem: A filename of the CA cert bundle to uses.
+                 You can specify this argument if you want to use a different
+                 CA cert bundle than the one used by botocore.
+    :type verify: bool or str
+    """
+
+    @apply_defaults
+    def __init__(
+            self,
+            query: str,
+            s3_bucket: str,
+            s3_key: str,
+            mysql_conn_id: str = 'mysql_default',
+            aws_conn_id: str = 'aws_default',
+            verify: Optional[Union[bool, str]] = None,
+            header: Optional[bool] = False,
+            index: Optional[bool] = False,
+            *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+        self.query = query
+        self.s3_bucket = s3_bucket
+        self.s3_key = s3_key
+        self.mysql_conn_id = mysql_conn_id
+        self.aws_conn_id = aws_conn_id
+        self.verify = verify
+        self.header = header
+        self.index = index
+
+    def fill_na_with_none(self, series):
+        """Replace NaN values with None"""
+        return np.where(series.isnull(), None, series)
+
+    def fix_int_dtypes(self, df):
+        """
+        Mutate DataFrame to set dtypes for int columns containing NaN values."
+        """
+        for col in df:
+            if "float" in df[col].dtype.name and df[col].hasnans:
+                # inspect values to determine if dtype of non-null values is int or float
+                notna_series = df[col].dropna().values
+                if np.isclose(notna_series, notna_series.astype(int)).all():
+                    # set to dtype that retains integers and supports NaNs
+                    df[col] = self.fill_na_with_none(df[col]).astype(pd.Int64Dtype)
+
+    def execute(self, context, **kwargs):
+        self.hook = MySqlHook(mysql_conn_id=self.mysql_conn_id)
+        self.s3_conn = S3Hook(aws_conn_id=self.aws_conn_id, verify=self.verify)
+        data_df = self.hook.get_pandas_df(self.query)
+        self.log.info("Data from MySQL obtained")
+
+        self.fix_int_dtypes(data_df)
+        file_obj = StringIO()
```

Comment:
Wouldn't it be better to use `NamedTemporaryFile`? 
```python
with NamedTemporaryFile(suffix='.csv') as tmp_csv:
    df.to_file(tmp_csv)
    self.s3_conn.load_file_obj(file_obj=file_obj,
                                    key=self.s3_key,
                                    bucket_name=self.s3_bucket,
                                    **kwargs)
```
In this way the file will be deleted after upload, thus no side effect will occur. I am not sure if this code works so please test it.

Reply:
By writing to a buffer, no cleanup is needed after upload to S3.

Reply:
While it's true that writing to a buffer doesn't require any cleanup, could it be that creating a temporary file as @nuclearpinguin suggests is better for larger datasets?

Reply:
@JavierLopezT 
This is correct. I was mistaken and thought this comment was related to `test_mysql_to_s3_operator.py`.

Reply:
I have been testing this. Unfortunately (with the changes I had to make with pickle) this doesn't generate a readable csv. I have been trying to imitate the functionality of NamedTemporaryFile using:
```
with open('tmp_file.csv', 'w+') as tmp_csv:
    df.to_csv(tmp_csv)
    temporary_file_path = os.path.realpath(tmp_csv.name)
os.remove(temporary_file_path)
```
However, Ubuntu seems to not have the permissions to create the file, so I have tried to give permissions prior to the open with:
`os.chmod(os.path.dirname(os.path.abspath(__file__)), 0o777)`
And it shows that it has no permissions for executing the command. Having to set the permissions manually seems a very bad idea for me. So, unless you have any better idea I think I will have to come back to StringIO

Reply:
Why is the DataFrame pickled then written to a file with a csv extension? This would definitely explain why the csv is unreadable. 

https://github.com/apache/airflow/blob/69ff9f3dbf25ecefc627f95bbb75e3b9b9835b7c/airflow/operators/mysql_to_s3_operator.py#L108

Reply:
> Why is the DataFrame pickled then written to a file with a csv extension? This would definitely explain why the csv is unreadable.
> 
> https://github.com/apache/airflow/blob/69ff9f3dbf25ecefc627f95bbb75e3b9b9835b7c/airflow/operators/mysql_to_s3_operator.py#L108

Because if you put the code as @nuclearpinguin proposed originally you get the error:
`AttributeError: 'DataFrame' object has no attribute 'to_file'`
And if you do
```
with tempfile.NamedTemporaryFile(suffix='.csv') as tmp_csv:
    df.to_csv(tmp_csv)
```
Then you get the error:
`TypeError: a bytes-like object is required, not 'str'`
So the only way I was able to get rid of errors was using pickle in that way. Maybe there is another workaround that I am not aware of.

Reply:
Gotcha. This has been an issue for me in the past, as well. 

One caveat is that the to_csv kwarg `index` should be set to False to enforce compatibility when loading from S3 into Redshift / Athena / MySQL. I would recommend including a `pd_csv_kwargs` in the operator to allow for users to set these options as needed. 

```Python

# in operator 
pd_csv_kwargs = {"index": False}

# write to temp file
with tempfile.NamedTemporaryFile(mode="r+", suffix='.csv') as tmp_csv:
    tmp_csv.file.write(df.to_csv(**pd_csv_kwargs))
    tmp_csv.file.seek(0)
```

Reply:
Thank you very much for your help. Regarding the kwargs, I already had included both index and header in the __init__ as ```
header: Optional[bool] = False,
            index: Optional[bool] = False
```
But I indeed forgot to include them in the code. Do you recommend to change that to your pd_csv_kwargs?

Reply:
I think including the option is a good idea. The implementation is tricky to support index / header kwargs and any other supported by the pandas method. I have a custom operator that works similar to this and settled on an approach similar to what is below. The overall goal being to set `index=False` by default. 

```Python
def __init__(
    self,
    ...,
    index=False,
    pd_csv_kwargs=None
):

    if not pd_csv_kwargs:
        pd_csv_kwargs = {}

    if "index" not in pd_csv_kwargs:
        pd_csv_kwargs["index"] = index
```

Reply:
Just added your suggestion. Thanks!

Reply:
@JavierLopezT Happy to help! Great work on this one.